base/sparse/sparsematrix.jl

```diff
@@ -193,20 +193,51 @@ end
 copy(S::SparseMatrixCSC) =
     SparseMatrixCSC(S.m, S.n, copy(S.colptr), copy(S.rowval), copy(S.nzval))
 
-function copy!{TvA, TiA, TvB, TiB}(A::SparseMatrixCSC{TvA,TiA},
-                                   B::SparseMatrixCSC{TvB,TiB})
-    # If the two matrices have the same size then all the
-    # elements in A will be overwritten and we can simply copy the
-    # internal fields of B to A.
-    if size(A) == size(B)
-        copy!(A.colptr, B.colptr)
+function copy!(A::SparseMatrixCSC, B::SparseMatrixCSC)
+    # If the two matrices have the same length then all the
+    # elements in A will be overwritten.
+    if length(A) == length(B)
         resize!(A.nzval, length(B.nzval))
         resize!(A.rowval, length(B.rowval))
-        copy!(A.rowval, B.rowval)
-        copy!(A.nzval, B.nzval)
+        if size(A) == size(B)
+            # Simple case: we can simply copy the internal fields of B to A.
+            copy!(A.colptr, B.colptr)
+            copy!(A.rowval, B.rowval)
+        else
+            # This is like a "reshape B into A".
+            sparse_compute_reshaped_colptr_and_rowval(A.colptr, A.rowval, A.m, A.n, B.colptr, B.rowval, B.m, B.n)
+        end
     else
-        invoke(Base.copy!, Tuple{AbstractMatrix{TvA}, AbstractMatrix{TvB}}, A, B)
+        length(A) >= length(B) || throw(BoundsError())
+        lB = length(B)
+        nnzA = nnz(A)
+        nnzB = nnz(B)
+        # Up to which col, row, and ptr in rowval/nzval will A be overwritten?
+        lastmodcolA = div(lB - 1, A.m) + 1
+        lastmodrowA = mod(lB - 1, A.m) + 1
```

Comment:
There's also `divrem`, if it makes sense. Or just work with `lB - lastmodcolA*A.m`.